### PR TITLE
Extract more math methods and add to global namespace

### PR DESCRIPTION
**Pull request by dsmilkov**
- Consolidate `math.executeOp()` and `math.scope()` into `math.scope()`:
  - `math.scope(() => {})` denotes an unnamed scope.
  - `math.scope('name', () => {})` replaces `math.executeOp('name', () => {})` and denotes a named scope.

- Fixed memory leaks of migrated ops by wrapping each op in `ENV.math.scope(opName, () => {});`

- Methods extracted and added to the global namespace:
  - vectorTimesMatrix
  - outerProduct
  - matrixTimesVector
  - dotProduct
  - slice1D/2D/3D/4D
  - reverse1D/2D/3D/4D
  - concat1D/2D/3D/4D
  - batchNormalization2D/3D/4D
  - {avg,min,max}Pool
  - conv1d/2d
  - conv2dTranspose
  - depthwiseConv2D

---
This change is [Reviewable](https://reviewable.io/reviews/pair-code/deeplearnjs/575)

### CODE REVIEW

**Comment by nsthorat** [comment]
Reviewed 7 of 13 files at r1.
Review status: 7 of 13 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/environment.ts, line 185 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-L3OjObZZ4Z_xznBmIC8:-L3OjObZZ4Z_xznBmIC9:bwlmmko) ([raw file](https://github.com/pair-code/deeplearnjs/blob/d168f45b13c5613591acf6812e194619aa1862dd/src/environment.ts#L185)):*
> ```XML
>   }
> 
>   getBestBackend(): BackendType {
> ```

s/getBestBackend/getBestBackendType

---

*[src/index.ts, line 64 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-L3Oj_3o8KAlpq-S5Pcv:-L3Oj_3o8KAlpq-S5Pcw:b6vktsl) ([raw file](https://github.com/pair-code/deeplearnjs/blob/d168f45b13c5613591acf6812e194619aa1862dd/src/index.ts#L64)):*
> ```XML
> export {avgPool, maxPool, minPool} from './math/pool';
> export {reverse1D, reverse2D, reverse3D, reverse4D} from './math/reverse';
> export {slice1D, slice2D, slice3D, slice4D} from './math/slice';
> ```

can we export all of the top level ops under a single "export", like we do below in "Second level exports" by first importing the functions and then exporting them together? That will help us organize this file a little better so it's not so scattered.

---

*[src/math/math.ts, line 85 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-L3Ok28nYBuIg7R-WW9p:-L3Ok28nYBuIg7R-WW9q:bnqgh9v) ([raw file](https://github.com/pair-code/deeplearnjs/blob/d168f45b13c5613591acf6812e194619aa1862dd/src/math/math.ts#L85)):*
> ```XML
>   maxPool = pool.maxPool;
>   minPool = pool.minPool;
>   /** @deprecated */
> ```

why are gradient methods deprecated and others aren't (same for methods below)?

---

*[src/math/math.ts, line 174 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-L3OkIc1Y3FWU3wvMlum:-L3OkIc1Y3FWU3wvMlun:b47tbz6) ([raw file](https://github.com/pair-code/deeplearnjs/blob/d168f45b13c5613591acf6812e194619aa1862dd/src/math/math.ts#L174)):*
> ```XML
> 
>   /**
>    * Runs the operations in the provided scope function. After the function
> ```

merge these into 1 sentence, the important part is not that it runs, rather that it cleans up intermediates:

Something along the lines of:

"Executes the provided function and after it is executed, cleans up all intermediate `NDArray`s allocated by the function except those returned by the function."

---

*[src/math/math.ts, line 181 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-L3OlqRHxhfxxar0J7IO:-L3OlqRHxhfxxar0J7IP:b-7v7nyo) ([raw file](https://github.com/pair-code/deeplearnjs/blob/d168f45b13c5613591acf6812e194619aa1862dd/src/math/math.ts#L181)):*
> ```XML
>    * inside a `math.scope()` to prevent memory leaks.
>    *
>    * @param name The name of the scope, optional. If a name is provided,
> ```

update this jsdoc because it can take a function directly

---

*[src/math/math.ts, line 186 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-L3OktL03Scio1uc2lNK:-L3OktL03Scio1uc2lNL:b5an833) ([raw file](https://github.com/pair-code/deeplearnjs/blob/d168f45b13c5613591acf6812e194619aa1862dd/src/math/math.ts#L186)):*
> ```XML
>    * @param scopeFn The function to execute.
>    */
>   scope<T extends ScopeResult>(name: string|ScopeFn<T>, scopeFn?: ScopeFn<T>):
> ```

maybe rename "name" to nameOrScopeFn

---

*[src/math/math.ts, line 209 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-L3OlCjGEqMuCpzIr8kI:-L3OlCjGEqMuCpzIr8kJ:b4tp4sa) ([raw file](https://github.com/pair-code/deeplearnjs/blob/d168f45b13c5613591acf6812e194619aa1862dd/src/math/math.ts#L209)):*
> ```XML
>     }
>     const gradientsMode = false;
>     return this.engine.scope('scope', scopeFn, gradientsMode);
> ```

pass the name through instead of just 'scope'

---

*[src/math/math.ts, line 219 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-L3OmAGQDw7XYJaYXTEl:-L3OmAGQDw7XYJaYXTEm:b-nbzo30) ([raw file](https://github.com/pair-code/deeplearnjs/blob/d168f45b13c5613591acf6812e194619aa1862dd/src/math/math.ts#L219)):*
> ```XML
>    */
>   gradientsScope<T extends ScopeResult>(
>       scopeFn:
> ```

we should do the same thing here, and pass the name to engine.scope

also type "scopeFn" as ScopeFn

---

*[src/math/math.ts, line 232 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-L3Oma5uIq6qdjdBo0pU:-L3OmNxnwGYbHhOfnP1K:bhzehr8) ([raw file](https://github.com/pair-code/deeplearnjs/blob/d168f45b13c5613591acf6812e194619aa1862dd/src/math/math.ts#L232)):*
> ```XML
>    * as scope() without the need for a function closure.
>    */
>   startScope() {
> ```

Just so we have two brains thinking about this, I never added "name" to startScope here. Don't worry about it in this PR because we don't actually do anything with name yet inside the backendEngine.scope. I'll take care of this in my PR that does hierarchical logging with scope functions + kernels.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575)*

**Comment by nsthorat** [comment]
:lgtm_strong:

---

Review status: 7 of 13 files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-:-L3Ompy3C0Oet1HJySQv:b-jbsf8z)*

**Comment by nsthorat** [comment]
Really nice work!! Scope with a first argument as name is nice.

---

Review status: 7 of 13 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-:-L3Omr6GWv0A9Ar5iYnI:b-8synb0)*

**Comment by dsmilkov** [comment]
Review status: 7 of 13 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[src/environment.ts, line 185 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-L3OjObZZ4Z_xznBmIC8:-L3P0Uf93OzkUFt8qvYW:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/d168f45b13c5613591acf6812e194619aa1862dd/src/environment.ts#L185)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

s/getBestBackend/getBestBackendType
</blockquote></details>

Done.

---

*[src/index.ts, line 64 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-L3Oj_3o8KAlpq-S5Pcv:-L3P1rACMq2iZOqMs948:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/d168f45b13c5613591acf6812e194619aa1862dd/src/index.ts#L64)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

can we export all of the top level ops under a single "export", like we do below in "Second level exports" by first importing the functions and then exporting them together? That will help us organize this file a little better so it's not so scattered.
</blockquote></details>

Done.

---

*[src/math/math.ts, line 85 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-L3Ok28nYBuIg7R-WW9p:-L3P1sLeTN0_tavJF0kJ:be9ob4r) ([raw file](https://github.com/pair-code/deeplearnjs/blob/d168f45b13c5613591acf6812e194619aa1862dd/src/math/math.ts#L85)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

why are gradient methods deprecated and others aren't (same for methods below)?
</blockquote></details>

talked offline. users shouldn't call backprop methods directly. they can use math.gradient(() => {}).

---

*[src/math/math.ts, line 174 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-L3OkIc1Y3FWU3wvMlum:-L3P27kiyDLjx90ySiPc:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/d168f45b13c5613591acf6812e194619aa1862dd/src/math/math.ts#L174)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

merge these into 1 sentence, the important part is not that it runs, rather that it cleans up intermediates:

Something along the lines of:

"Executes the provided function and after it is executed, cleans up all intermediate `NDArray`s allocated by the function except those returned by the function."
</blockquote></details>

Done.

---

*[src/math/math.ts, line 181 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-L3OlqRHxhfxxar0J7IO:-L3P2S5lspL8vTk9OyFU:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/d168f45b13c5613591acf6812e194619aa1862dd/src/math/math.ts#L181)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

update this jsdoc because it can take a function directly
</blockquote></details>

Done.

---

*[src/math/math.ts, line 186 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-L3OktL03Scio1uc2lNK:-L3P2Un7MhZCMWV2LN30:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/d168f45b13c5613591acf6812e194619aa1862dd/src/math/math.ts#L186)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

maybe rename "name" to nameOrScopeFn
</blockquote></details>

Done.

---

*[src/math/math.ts, line 209 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-L3OlCjGEqMuCpzIr8kI:-L3P2nXtcZrrC8Ob5r9v:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/d168f45b13c5613591acf6812e194619aa1862dd/src/math/math.ts#L209)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

pass the name through instead of just 'scope'
</blockquote></details>

Done.

---

*[src/math/math.ts, line 219 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-L3OmAGQDw7XYJaYXTEl:-L3P4dXqucq6Jn1P8MCc:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/d168f45b13c5613591acf6812e194619aa1862dd/src/math/math.ts#L219)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

we should do the same thing here, and pass the name to engine.scope

also type "scopeFn" as ScopeFn
</blockquote></details>

Done.

---

*[src/math/math.ts, line 232 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575#-L3Oma5uIq6qdjdBo0pU:-L3P2tFOwerrUMLf8Qdr:bb6sh2i) ([raw file](https://github.com/pair-code/deeplearnjs/blob/d168f45b13c5613591acf6812e194619aa1862dd/src/math/math.ts#L232)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

Just so we have two brains thinking about this, I never added "name" to startScope here. Don't worry about it in this PR because we don't actually do anything with name yet inside the backendEngine.scope. I'll take care of this in my PR that does hierarchical logging with scope functions + kernels.
</blockquote></details>

Ack.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/575)*
